pkg/virtual/initializingworkspaces/builder/build.go

```diff
@@ -0,0 +1,184 @@
+	if !strings.HasSuffix(rootPathPrefix, "/") {
+		rootPathPrefix += "/"
+	}
+
+	return &virtualworkspacesdynamic.DynamicVirtualWorkspace{
+		Name: initializingworkspaces.VirtualWorkspaceName,
+		RootPathResolver: func(urlPath string, requestContext context.Context) (accepted bool, prefixToStrip string, completedContext context.Context) {
+			debug.PrintStack()
+
+			completedContext = requestContext
+			if !strings.HasPrefix(urlPath, rootPathPrefix) {
+				return
+			}
+			withoutRootPathPrefix := strings.TrimPrefix(urlPath, rootPathPrefix)
+
+			// Incoming requests to this virtual workspace will look like:
+			//  /services/initializingworkspaces/<initializer>/clusters/*/apis/workload.kcp.dev/v1alpha1/workloadclusters
+			//                                  └───────────┐
+			// Where the withoutRootPathPrefix starts here: ┘
+			parts := strings.SplitN(withoutRootPathPrefix, "/", 2)
+			if len(parts) < 2 {
+				return
+			}
+
+			initializerName := parts[0]
+			if initializerName == "" {
+				return
+			}
+
+			realPath := "/" + parts[1]
+
+			//  /services/initializingworkspaces/<initializer>/clusters/*/apis/workload.kcp.dev/v1alpha1/workloadclusters
+			//                  ┌─────────────────────────────┘
+			// We are now here: ┘
+			// Now, we parse out the logical cluster and validate that only wildcard requests are allowed.
+			cluster := genericapirequest.Cluster{Name: logicalcluster.Wildcard, Wildcard: true}
+			if strings.HasPrefix(realPath, "/clusters/") {
+				withoutClustersPrefix := strings.TrimPrefix(realPath, "/clusters/")
+				parts = strings.SplitN(withoutClustersPrefix, "/", 2)
+				lclusterName := parts[0]
+				realPath = "/"
+				if len(parts) > 1 {
+					realPath += parts[1]
+				}
+				if lclusterName != "*" {
+					return false, "", nil
+					// TODO: how do we signal that this is an error?
+				}
+			}
+
+			completedContext = genericapirequest.WithCluster(requestContext, cluster)
+			completedContext = dynamiccontext.WithAPIDomainKey(completedContext, dynamiccontext.APIDomainKey(initializerName))
+			prefixToStrip = strings.TrimSuffix(urlPath, realPath)
+			accepted = true
+			return
+		},
+		Ready: func() error {
+			return nil
+		},
+		BootstrapAPISetManagement: func(mainConfig genericapiserver.CompletedConfig) (apidefinition.APIDefinitionSetGetter, error) {
+			return &apiSetRetriever{
+				config:               mainConfig,
+				dynamicClusterClient: dynamicClusterClient,
+			}, nil
+		},
+	}
+}
+
+type apiSetRetriever struct {
+	config               genericapiserver.CompletedConfig
+	dynamicClusterClient dynamic.ClusterInterface
+}
+
+func (a *apiSetRetriever) GetAPIDefinitionSet(ctx context.Context, key dynamiccontext.APIDomainKey) (apis apidefinition.APIDefinitionSet, apisExist bool, err error) {
+	initializerName := string(key)
+
+	specs, err := internalapis.CreateAPIResourceSchemas(
+		[]*runtime.Scheme{scheme},
+		[]common.GetOpenAPIDefinitions{kcpopenapi.GetOpenAPIDefinitions},
```

Comment:
I don't think calling this function makes sense here. Instead us a constructor that turns a CRD into an APIResourceSchema. That's mostly trivial because they have same structure and share lots of types. @ncdc had built a cli tool and might be able to point you to that code.